File: src/org/webfabric/sitemesh/StringTemplateDecoratorServlet.java
package org.webfabric.sitemesh;

import com.opensymphony.module.sitemesh.HTMLPage;
import com.opensymphony.module.sitemesh.RequestConstants;
import org.antlr.stringtemplate.StringTemplate;
import org.antlr.stringtemplate.StringTemplateGroup;
import org.antlr.stringtemplate.language.DefaultTemplateLexer;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

import fj.data.List;
import static fj.data.List.list;
import fj.F;

public class StringTemplateDecoratorServlet extends HttpServlet {
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        StringTemplate template = getTemplate(request);
        template.setAttribute("request", request);
        template.setAttribute("response", response);

        HTMLPage html = getPage(request);

        StringTemplateDecorator templateDecorator = new StringTemplateDecorator(template);
        templateDecorator.Decorate(html, response.getWriter());
    }

    private StringTemplate getTemplate(final HttpServletRequest request) {
        List<Path> possiblePaths = list(new OriginalServletPath(request),
                new OriginalPathInfo(request),
                new ServletPath(request),
                new PathInfo(request));

        Path validPath = possiblePaths.dropWhile(new F<Path, Boolean>() {
            public Boolean f(Path path1) {
                return path1.value() == null;
            }
        }).head();
        String name = removeExtension(validPath.value());

        StringTemplateGroup groups = new StringTemplateGroup("decorators", getRootPath(), DefaultTemplateLexer.class);
        return groups.getInstanceOf(name);
    }

    private String getRootPath() {
        return getServletContext().getRealPath("");
    }

    private String removeExtension(String path) {
        return path.replace(".st", "");
    }

    private HTMLPage getPage(HttpServletRequest request) {
        return (HTMLPage) request.getAttribute(RequestConstants.PAGE);
    }

}

File: src/org/webfabric/sitemesh/StringTemplateDecoratorServlet.scala
package org.webfabric.sitemesh

import javax.servlet.http._
import org.webfabric.stringtemplate._
import org.webfabric.servlet._
import org.webfabric.io.{Url}
import org.antlr.stringtemplate.{StringTemplate}

class StringTemplateDecoratorServlet extends HttpServlet {
  val contentProvider = new AutoDetectingContentProvider()
  override def doPost(request: HttpServletRequest, response: HttpServletResponse) = {
    handle(request, response)
  }

  override def doGet(request: HttpServletRequest, response: HttpServletResponse) = {
    handle(request, response)
  }

  def handle(request: HttpServletRequest, response: HttpServletResponse) {
    val templateDecorator = new StringTemplateDecorator(getTemplate(request))
    templateDecorator.setBase(ContextPath(request))
    templateDecorator.setQueryString(QueryString(request))
    templateDecorator.setInclude(getPageMap(request, response))

    contentProvider.getContent(request) match {
      case Some(content) => templateDecorator.setContent(content)
      case None =>
    }

    templateDecorator.writeTo(response.getWriter())
  }

  def getTemplate(request: HttpServletRequest): StringTemplate = {
    val possiblePaths = List[String](OriginalServletPath(request),
      OriginalPathInfo(request),
      ServletPath(request),
      PathInfo(request))

    val StringTemplateDecoratorServlet.TemplateName(path, name) = possiblePaths.find(_ != null).get
    val groups = new UrlStringTemplateGroup("decorators", getBaseUrl(path))
    return groups.getInstanceOf(name)
  }

  def getBaseUrl(path:String): Url = {
    getServletContext().getResource(path)
  }

  def getPageMap(request: HttpServletRequest, response: HttpServletResponse):PageMap = {
    val loaders = new PageLoaders(new UrlPageLoader, new v2.ServletPageLoader(request, response, getServletConfig))
    new PageMap(loaders)
  }
}

object StringTemplateDecoratorServlet {
  val TemplateName = """(?:(.+)/)?(.+)\.st$""".r
}